### PR TITLE
Allowing for non-default-constructible component types

### LINKED ISSUE #2848: Components have to be default constructible
Here is another problem related to types without an accessible default constructor.
`tests/unit/component/new_.cpp` does not compile after it's changed like this,
```c++
diff --git a/tests/unit/component/new_.cpp b/tests/unit/component/new_.cpp
index 16185e1fa3..341e858dd4 100644
--- a/tests/unit/component/new_.cpp
+++ b/tests/unit/component/new_.cpp
@@ -15,8 +15,12 @@
 ///////////////////////////////////////////////////////////////////////////////
 struct test_server : hpx::components::simple_component_base<test_server>
 {
+    test_server() = delete;
+    test_server(int i) : i(i) {}
     hpx::id_type call() const { return hpx::find_here(); }
 
+    int i;
+
     HPX_DEFINE_COMPONENT_ACTION(test_server, call);
 };                                                                                                                                                                                       
                                                                                                                                                                                          
@@ -49,7 +53,7 @@ void test_create_single_instance()                                                                                                                                      
     // make sure created objects live on locality they are supposed to be                                                                                                                
     for (hpx::id_type const& loc: hpx::find_all_localities())                                                                                                                            
     {                                                                                                                                                                                    
-        hpx::id_type id = hpx::new_<test_server>(loc).get();                                                                                                                             
+        hpx::id_type id = hpx::new_<test_server>(loc, 42).get();                                                                                                                         
         HPX_TEST(hpx::async<call_action>(id).get() == loc);                                                                                                                              
     }                                                                                                                                                                                    
                                                                                                                                                                                          
@@ -60,7 +64,7 @@ void test_create_single_instance()                                                                                                                                      
     }                                                                                                                                                                                    
                                                                                                                                                                                          
     // make sure distribution policy is properly used                                                                                                                                    
-    hpx::id_type id = hpx::new_<test_server>(hpx::default_layout).get();                                                                                                                 
+    hpx::id_type id = hpx::new_<test_server>(hpx::default_layout, 42).get();                                                                                                             
     HPX_TEST(hpx::async<call_action>(id).get() == hpx::find_here());
 
     test_client t2 = hpx::new_<test_client>(hpx::default_layout);
@@ -68,7 +72,7 @@ void test_create_single_instance()
 
     for (hpx::id_type const& loc: hpx::find_all_localities())
     {
-        hpx::id_type id = hpx::new_<test_server>(hpx::default_layout(loc)).get();
+        hpx::id_type id = hpx::new_<test_server>(hpx::default_layout(loc), 42).get();
         HPX_TEST(hpx::async<call_action>(id).get() == loc);
     }
 
@@ -85,7 +89,7 @@ void test_create_multiple_instances()
     // make sure created objects live on locality they are supposed to be
     for (hpx::id_type const& loc: hpx::find_all_localities())
     {
-        std::vector<hpx::id_type> ids = hpx::new_<test_server[]>(loc, 10).get();
+        std::vector<hpx::id_type> ids = hpx::new_<test_server[]>(loc, 10, 42).get();
         HPX_TEST_EQ(ids.size(), std::size_t(10));
 
         for (hpx::id_type const& id: ids)
@@ -96,7 +100,7 @@ void test_create_multiple_instances()
 
     for (hpx::id_type const& loc: hpx::find_all_localities())
     {
-        std::vector<test_client> ids = hpx::new_<test_client[]>(loc, 10).get();
+        std::vector<test_client> ids = hpx::new_<test_client[]>(loc, 10, 42).get();
         HPX_TEST_EQ(ids.size(), std::size_t(10));
 
         for (test_client const& c: ids)
@@ -107,7 +111,7 @@ void test_create_multiple_instances()
 
     // make sure distribution policy is properly used
     std::vector<hpx::id_type> ids =
-        hpx::new_<test_server[]>(hpx::default_layout, 10).get();
+        hpx::new_<test_server[]>(hpx::default_layout, 10, 42).get();
     HPX_TEST_EQ(ids.size(), std::size_t(10));
     for (hpx::id_type const& id: ids)
     {
@@ -115,7 +119,7 @@ void test_create_multiple_instances()
     }
 
     std::vector<test_client> clients =
-        hpx::new_<test_client[]>(hpx::default_layout, 10).get();
+        hpx::new_<test_client[]>(hpx::default_layout, 10, 42).get();
     HPX_TEST_EQ(clients.size(), std::size_t(10));
     for (test_client const& c: clients)
     {
@@ -125,7 +129,7 @@ void test_create_multiple_instances()
     for (hpx::id_type const& loc: hpx::find_all_localities())
     {
         std::vector<hpx::id_type> ids =
-            hpx::new_<test_server[]>(hpx::default_layout(loc), 10).get();
+            hpx::new_<test_server[]>(hpx::default_layout(loc), 10, 42).get();
         HPX_TEST_EQ(ids.size(), std::size_t(10));
 
         for (hpx::id_type const& id: ids)
@@ -137,7 +141,7 @@ void test_create_multiple_instances()
     for (hpx::id_type const& loc: hpx::find_all_localities())
     {
         std::vector<test_client> ids =
-            hpx::new_<test_client[]>(hpx::default_layout(loc), 10).get();
+            hpx::new_<test_client[]>(hpx::default_layout(loc), 10, 42).get();
         HPX_TEST_EQ(ids.size(), std::size_t(10));
 
         for (test_client const& c: ids)
```
```
hpx.git/hpx/runtime/components/server/simple_component_base.hpp:304:64: error: use of deleted function ‘test_server::test_server()’
             return static_cast<component_type*>(new Component()); //-V572
                                                                ^
hpx.git/tests/unit/component/new_.cpp:18:5: note: declared here
     test_server() = delete;
```

### PR DESCRIPTION
**Pull request by hkaiser**
- this fixes #2848

@krivenko: please verify this solves the problem you reported the other day,

### CODE REVIEW

**Comment by krivenko** [comment]
@hkaiser
The `new_2848` test compiles and passes.
My application code compiles as well with default constructors removed from components.
Thanks a lot!

File: hpx/runtime/components/component_factory.hpp [hunk 1]
```diff
@@ -166,9 +168,25 @@ namespace hpx { namespace components
...
+            return create<Component>(count,
+                std::is_default_constructible<
+                    typename Component::type_holder>());
+        }
+
+        template <typename Component_>
+        naming::gid_type create(std::size_t count, std::false_type)
+        {
+            // shouldn't ever be called for non-default-constructible types
+            HPX_ASSERT(false);
```
Comment (sithhell) [inline hpx/runtime/components/component_factory.hpp]: Why do we have to resort to a runtime error here?

File: hpx/runtime/components/derived_component_factory.hpp [hunk 2]
```diff
@@ -153,6 +155,22 @@ namespace hpx { namespace components
...
+            return create<Component>(count,
+                std::is_default_constructible<
+                    typename Component::type_holder>());
+        }
+
+        template <typename Component_>
+        naming::gid_type create(std::size_t count, std::false_type)
+        {
+            // shouldn't ever be called for non-default-constructible types
+            HPX_ASSERT(false);
```
Comment (sithhell) [inline hpx/runtime/components/derived_component_factory.hpp]: same as above

File: hpx/runtime/components/component_factory.hpp (hunk 1 above)
Comment (hkaiser) [inline hpx/runtime/components/component_factory.hpp]: What do you suggest? A static assert would trigger for any non-default constructible type.

Comment (sithhell) [inline hpx/runtime/components/component_factory.hpp]: Only if we call new_ without arguments, no?

Comment (hkaiser) [inline hpx/runtime/components/component_factory.hpp]: No, always - this function is called by `create(size_t)` which is a virtual function.

Comment (sithhell) [inline hpx/runtime/components/component_factory.hpp]: ah, right, too bad :/

**Pull request merged**